libs/platform/src/lib/components/form/checkbox/checkbox.component.ts

```diff
@@ -0,0 +1,226 @@
+    encapsulation: ViewEncapsulation.None
+})
+export class CheckboxComponent extends BaseInput implements AfterViewInit {
+    /** set to true if binary checkbox */
+    @Input()
+    isBinary: boolean = false;
+
+    /** Sets label for checkbox. */
+    @Input()
+    label: string;
+
+    /**
+     * Attached to the aria-label attribute of the host element. In most cases, aria-labelledby will
+     * take precedence so this may be omitted.
+     */
+    @Input('aria-label')
+    ariaLabel: string = '';
+
+    /**
+     * Users can specify the `aria-labelledby` attribute which will be forwarded to the input element
+     */
+    @Input('aria-labelledby')
+    ariaLabelledby: string | null = null;
+
+    /** indeterminate state */
+    @Input()
+    tristate: boolean = false;
+
+    /** indeterminate state can be selected  */
+    @Input()
+    tristateSelectable: boolean = true;
+
+    /** value for checkbox, when it is not binary */
+    @Input()
+    get value(): any {
+        return super.getValue();
+    }
+    set value(selectValue: any) {
+        if (selectValue !== this.value) {
+            // baseInput class is calling this.onchange(value),
+            // which causing Expression change error, so doing all steps here.
+            this._value = selectValue;
+            this.stateChanges.next('writeValue');
+            this._cd.markForCheck();
+        }
+    }
+
+    /** Emitting checkbox change event */
+    @Output()
+    change = new EventEmitter();
+
+    /** @hidden
+     * tracking checkbox current value
+     */
+    public checkboxValue: any;
+
+    /* @hidden
+     * reference of child checkbox implementation
+     */
+    @ViewChild(CoreCheckboxComponent)
+    private corecheckbox: CoreCheckboxComponent;
+
+    /* @hidden
+     * stores  formControl values
+     */
+    private multiSelectModel: any;
+
+    constructor(
+        protected _changeDetector: ChangeDetectorRef,
+        @Optional() @Self() public ngControl: NgControl,
+        @Optional() @Self() public ngForm: NgForm
+    ) {
+        super(_changeDetector, ngControl, ngForm);
+    }
+
+    /** ControlvalueAccessor */
+    writeValue(value: any): void {
+        if (value) {
+            // formcontrol is expected as array or binary
+            this._setCoreCheckboxControl(value);
+        }
+    }
+
+    /** child checkbox is initialized here */
+    ngAfterViewInit(): void {
+        if (!this.isBinary) {
+            // updating checkbox values property for this custom checkbox
+            this.corecheckbox.values.trueValue = this.value;
+            this.corecheckbox.values.falseValue = undefined;
+
+            // set core checkbox control value based on platform checkbox control value
+            if (this.value && this.multiSelectModel && this.multiSelectModel.includes(this.value)) {
+                this.checkboxValue = this.value;
+            } else {
+                if (this.tristate && this.multiSelectModel.includes(null)) {
```

Comment:
I'm unsure about this if statement, it implies that we have a tristate checkbox(es) as part of a multi select group. To me "tristate" and "multi-select" are mutually exclusive - when you are checking an item from a multi select group, that item is either "in" or "out" of the select group, it can't be in a third state. This if statement kind of illustrates that, since if you have multiple tristate checkboxes in a multi select group, they may all have the same "null" value.

Can anyone provide an example where we need a tristate checkbox in a multi select?

Reply:
have changed the logic for tristate after discussion.